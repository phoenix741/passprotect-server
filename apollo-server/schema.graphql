scalar Date

# Encrypted content
type EncryptedContent {
  # Content encrypted with salt
  content: String!
  # Authentification tag
  authTag: String!
}

# Encrypted content
input EncryptedContentInput {
  # Content encrypted with salt
  content: String!
  # Authentification tag
  authTag: String!
}

type Error {
  fieldName: String,
  message: String
}

type Errors {
  errors: [Error]!
}

# Master key encrypted for the user
type EncryptedMasterKey {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContent!
}

# Master key encrypted for the user
input EncryptedMasterKeyInput {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContentInput!
}

# User defined in the application
type User {
  # Login of the user
  _id: ID!,
  # Encrypted master key
  encryption: EncryptedMasterKey!,
  # Creation date of the user
  creationAt: Date,
  # Role of the user (can be user or admin)
  role: String
  # Wallet lines of the user.
  lines: [WalletLine]!
}

input RegistrationUser {
  # Login of the user
  _id: ID!,
  # Password used for registration
  password: String!
  # Encrypted master key
  encryption: EncryptedMasterKeyInput!,
}

union RegisterUserResult = User | Errors

# Wallet line content encrypted
type EncryptedWalletLine {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContent!
}

# Wallet line content encrypted
input EncryptedWalletLineInput {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContentInput!
}

# Wallet
type WalletLine {
  # Id of the line
  _id: ID!
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLine!,
  # Owner of the list
  user: User!,
  # Last updated date of the line
  updatedAt: Date,
  # Revision (modification count) of the line
  _rev: Int
}

# Wallet line used as input
input WalletLineInput {
  # Id of the line, if not filled application create a new line
  _id: ID
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLineInput!,
  # Revision (modification count) of the line, used to ensure modification of the last line
  _rev: Int
}

union CreateUpdateLineResult = WalletLine | Errors

# Information needed to connect to the server and retreive object from database.
input ConnectionInformationInput {
  # Login of the user
  username: String!,
  # Password used for connection
  password: String!
}

# Result of the connection
type ConnectionInformation {
  # Token to be used in connection information
  token: String,
  # The connected user
  user: User!
}

union CreateSessionResult = ConnectionInformation | Errors

# Wallet transaction, used to synchronize lines between devices
type WalletTransaction {
  # Id of the transaction
  _id: ID!,
  # Type of the object modified (only line)
  type: String!,
  # Line
  line: WalletLine,
  # Owner of the transaction
  user: User!,
  # The line before the modification
  before: WalletLine,
  # The line after the modification
  after: WalletLine,
  # The modification date
  updatedAt: Date!,
  # A SHA-512 calculate on the modification.
  sha512: String!
}

type RootQuery {
  # Request a user
  user(id: ID!): User,

    # Request the list of lines from the connected users.
  lines: [WalletLine],

  # Request a wallet line
  line(id: ID!): WalletLine,

	# Request all groups available on lines
	groups: [String!]

  # Request the current session
  session: User

  # Request all transactions that happen since the last synchronisation date.
  transactions(earliest: Date!): [WalletTransaction]
}

type RootMutation {
  registerUser(input: RegistrationUser!): RegisterUserResult

  createUpdateLine(input: WalletLineInput!): CreateUpdateLineResult
  removeLine(id: ID): Errors

  createSession(input: ConnectionInformationInput!): CreateSessionResult
}

type RootSubscription {
  transactionAdded: WalletTransaction!
}

schema {
  query: RootQuery
  mutation: RootMutation
  subscription: RootSubscription
}
