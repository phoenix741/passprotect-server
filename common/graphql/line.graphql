# Wallet line content encrypted
type EncryptedWalletLine {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContent!
}

# Wallet line content encrypted
input EncryptedWalletLineInput {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContentInput!
}

# Wallet
type WalletLine {
  # Id of the line
  _id: ID!
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLine!,
  # Owner of the list
  user: User!,
  # Last updated date of the line
  updatedAt: Date,
  # Revision (modification count) of the line
  _rev: Int
}

# Wallet line used as input
input WalletLineInput {
  # Id of the line, if not filled application create a new line
  _id: ID
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLineInput!,
  # Revision (modification count) of the line, used to ensure modification of the last line
  _rev: Int
}

extend type RootQuery {
  # Request the list of lines from the connected users.
  lines: [WalletLine],

  # Request a wallet line
  line(id: ID!): WalletLine,

	# Request all groups available on lines
	groups: [String!]
}

union CreateUpdateLineResult = WalletLine | Errors

extend type RootMutation {
  createUpdateLine(input: WalletLineInput!): CreateUpdateLineResult
  removeLine(id: ID): Errors
}
