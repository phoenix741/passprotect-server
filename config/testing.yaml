config:
    mongodb:
        host: 'mongodb://localhost:27017/passprotect'
        user: ~
        password: ~
        options: ~
