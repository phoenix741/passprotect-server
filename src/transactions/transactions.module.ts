import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { TransactionSchema } from './schemas/transaction.schema';
import { TransactionsService } from './transactions.service';
import { SharedModule } from '../shared/shared.module';

@Module({
  imports: [
    SharedModule,
    MongooseModule.forFeature([{ name: 'Transaction', schema: TransactionSchema }]),
  ],
  providers: [TransactionsService],
  exports: [TransactionsService],
})
export class TransactionsModule {}
