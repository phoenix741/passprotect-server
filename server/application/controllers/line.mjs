import {omit, pick, includes, isString, isEmpty} from 'lodash'
import expressPromiseRouter from 'express-promise-router'
import i18n from 'i18next'
import fs from 'fs'
import path from 'path'
import debug from 'debug'
import {authenticate, permission, checkPermission} from '../utils/passport'
import {getLines, getLine, removeLine, saveLine} from '../services/line'
import {getGroups} from '../services/group'

const log = debug('App:Controllers:Line')

const router = expressPromiseRouter()
export default [authenticate(), permission(), router]

log('Load line type definition')
export const typeDefs = [
  fs.readFileSync(path.join(__dirname, '..', '..', '..', 'common', 'graphql', 'line.graphql'), 'utf-8')
]

export const resolvers = {
  RootQuery: {
    lines (obj, args, {user}) {
      return checkPermission(user).then(() => getLines(user)).map(line => checkPermission(user, [], line.user).return(line)).map(filterLine)
    },

    line (obj, {id}, {user}) {
      return getLine(id).tap(line => checkPermission(user, [], line.user)).then(filterLine)
    },

    groups (obj, args, {user}) {
      return checkPermission(user).then(() => getGroups(user))
    }
  },
  User: {
    lines (obj, args, {user}) {
      return getLines(obj._id).map(line => checkPermission(user, [], line.user).return(line)).map(filterLine)
    }
  },

  RootMutation: {
    createUpdateLine (obj, {input}, {user}) {
      let checkPermissionPromise = Promise.resolve(input)
      if (input._id) {
        checkPermissionPromise = getLine(input._id)
          .tap(line => checkPermission(user, [], line.user))
          .then(line => {
            input._id = line._id
            return input
          })
      }

      input.user = user._id
      return checkPermissionPromise
        .then(input => sanitizeInput(input))
        .then(data => saveLine(data))
        .then(filterLine)
        .catch(parseErrors)
    },

    removeLine (obj, {id}, {user}) {
      return getLine(id)
        .tap(line => checkPermission(user, [], line.user))
        .then(() => removeLine(id))
        .then(() => ({errors: []}))
        .catch(parseErrors)
    }
  },

  CreateUpdateLineResult: {
    __resolveType (obj) {
      if (obj.errors) {
        return 'Errors'
      }
      return 'WalletLine'
    }
  }
}

/**
 * @apiDefine text Informations to encrypt (client side) for text
 */
/**
 * @apiDefine password Informations to encrypt (client side) for password
 */
/**
 * @apiDefine card Informations to encrypt (client side) for payment card
 */

/**
 * @apiDefine UpdateLine
 *
 * @apiParam {String="text","card","password"} type type of line
 * @apiParam {String} label label of the line
 * @apiParam {Object} encryption Informations used for encrypted data
 * @apiParam {Object} encryption.informations Encrypted informations
 * @apiParam {String} encryption.informations.authTag Authentification tag of the encrypted information
 * @apiParam {String} encryption.informations.content Encrypted informations
 * @apiParam {String} encryption.salt Salt used to encrypt/decrypt informations
 *
 * @apiParam (text) {String} text Text that will be encrypted
 * @apiParam (text) {String} notes Custom note added to the line
 *
 * @apiParam (password) {String} username Username associated to the pasword
 * @apiParam (password) {String} password The password
 * @apiParam (password) {String} siteUrl The site where the password is used
 * @apiParam (password) {String} notes Custom note added to the line
 *
 * @apiParam (card) {String="VISA","Carte bleu","Mastercard","American Express","Maestro"} type Type of card
 * @apiParam (card) {String} nameOnCard Name of the person on the card
 * @apiParam (card) {String} cardNumber The card number
 * @apiParam (card) {String} cvv The code in the back on the card
 * @apiParam (card) {String} expiry The expiration date
 * @apiParam (card) {String} code The code to use to pay with the card
 * @apiParam (card) {String} notes Custom note added to the line
 *
 * @apiParamExample {json} Line:
 *   {
 *     "type": "text",
 *     "label": "My custom text",
 *     "encryption": {
 * )      "salt": "83b01169a90fe764",
 *       "informations": {
 *         "content": "d13a32829c2547b8bca6da5bb568d3",
 *         "authTag": "db47515bdcb7643484d60189b7ad4e33"
 *       }
 *     }
 *   }
 */

/**
 * @apiDefine ResponseLineComplete
 *
 * @apiSuccess {String} _id Id of the line
 * @apiSuccess {String} type type of line
 * @apiSuccess {String} label label of the line
 * @apiSuccess {Object} encryption Informations used for encrypted data
 * @apiSuccess {Object} encryption.informations Encrypted informations
 * @apiSuccess {String} encryption.informations.content Encrypted content of the line
 * @apiSuccess {String} encryption.informations.authTag Authentification tag of the encrypted informations
 * @apiSuccess {String} encryption.salt Salt used to encrypt informations
 *
 *
 * @apiSuccessExample {json} Success text:
 *   HTTP/1.1 200 OK
 *     {
 *       "_id": "57685936f4495bc62c4d28a3",
 *       "type": "text",
 *       "label": "My custom text",
 *       "encryption": {
 *         "salt": "83b01169a90fe764",
 *         "informations": {
 *           "content": "d13a32829c2547b8bca6da5bb568d3",
 *           "authTag": "db47515bdcb7643484d60189b7ad4e33"
 *         }
 *       }
 *     }
 */

router.route('')
  /**
   * @api {get} /api/lines Request all lines
   *
   * @apiDescription Request the list of lines from the connected users. Only the _id, type, and label of the line is
   * given. No encrypted information are send in this method.
   *
   * @apiName GetLines
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiParam {Number} [offset=0] Offset for pagination
   * @apiParam {Number} [limit=10000] Limit of items used for pagination
   *
   * @apiSuccess {String} _id Id of the line
   * @apiSuccess {String} type type of line
   * @apiSuccess {String} label label of the line
   *
   * @apiSuccessExample Success:
   *     HTTP/1.1 200 OK
   *     [
   *       {
   *         "_id": "57685936f4495bc62c4d28a3",
   *         "type": "card",
   *         "label": "My credit card"
   *       },
   *       {
   *         "_id": "57685936f4495bc62c4d28a4",
   *         "type": "password",
   *         "label": "My password"
   *       },
   *     ]
   *
   * @apiError Unauthorized Only the connected user can access to list of lines
   * @apiError Forbidden Your are not allowed to access to this API
   *
   * @apiErrorExample {json} Unauthorized:
   *     HTTP/1.1 401 Unauthorized
   *
   * @apiErrorExample {json} Forbidden:
   *     HTTP/1.1 403 Forbidden
   */
  .get((req, res) => {
    return getLines(req.user).map(filterLine).then(res.json.bind(res))
  })

  /**
   * @api {post} /api/lines Create new line
   * @apiName PostLines
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiUse UpdateLine
   * @apiUse ResponseLineComplete
   *
   * @apiError Unauthorized Only the connected user can access to list of lines
   *
   * @apiErrorExample {json} Unauthorized:
   *     HTTP/1.1 401 Unauthorized
   */
  .post((req, res) => {
    const data = omit(req.body, '_id', 'user')

    return checkAndSaveLine(data, req, res)
  })

router.param('lineId', (req, res, next, lineId) => {
  getLine(lineId, req.user)
    .then(line => {
      // Can't access or update lines that isn't associated to the account of the user
      if (line.user !== req.user._id) {
        const notAuthorizedError = new Error('Not authorized to access to thhis subscriptions')
        notAuthorizedError.status = 403
        return Promise.reject(notAuthorizedError)
      }

      req.line = line
      return next()
    })
    .catch(next)
})

/**
 * @api {get} /api/lines/:lineId Request line
 * @apiName GetLine
 * @apiGroup Line
 * @apiPermission ROLE_USER
 *
 * @apiParam {String} lineId The id of the line
 *
 * @apiUse ResponseLineComplete
 *
 * @apiError Unauthorized Only the connected user can access to list of lines
 * @apiError Forbidden Your are not allowed to access to this API
 *
 * @apiErrorExample {json} Unauthorized:
 *     HTTP/1.1 401 Unauthorized
 *
 * @apiErrorExample {json} Forbidden:
 *     HTTP/1.1 403 Forbidden
 */
router.route('/:lineId')
  .get((req, res) => res.json(filterLine(req.line)))
  /**
   * @api {put} /api/lines/:lineId Update the line
   * @apiName PutLine
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiUse UpdateLine
   * @apiUse ResponseLineComplete
   */
  .put((req, res) => {
    // Force the id of the line to ensure that the user doesn't try to change it
    const data = omit(req.body, '_id')
    data._id = req.line._id

    return checkAndSaveLine(data, req, res)
  })
  /**
   * @api {delete} /api/lines/:lineId Delete the line
   * @apiName DeleteLine
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiSuccessExample {json} Success-Response:
   *     HTTP/1.1 204 OK
   */
  .delete((req, res) => removeLine(req.line._id).then(() => res.status(204).send()))

/**
 * Filter information that the user is authorized to view in the response
 * @param {Object} line The response
 * @returns {Object} The filtered object.
 */
function filterLine (line) {
  return pick(line, '_id', 'user', 'type', 'label', 'group', 'encryption', 'updatedAt', '_rev')
}

function sanitizeInput (input) {
  const data = pick(input, '_id', 'user', 'type', 'label', 'group', 'encryption', '_rev')

  const validationError = new Error()
  validationError.status = 400

  const typeChoices = ['card', 'password', 'text']
  if (!includes(typeChoices, data.type)) {
    validationError.message = i18n.t('error:line.400.type', {choices: typeChoices.join(', ')})
    return Promise.reject(validationError)
  }

  if (!isString(data.label) || isEmpty(data.label)) {
    validationError.message = i18n.t('error:line.400.label')
    return Promise.reject(validationError)
  }

  return Promise.resolve(data)
}

/**
 * Create or modify a line given by the request.
 * @param {Object} data Object to create or modify
 * @param {Object} req Request use to read the user
 * @param {Object} res Response used to send the response.
 * @returns {Promise} Promise that the line will be saved.
 */
function checkAndSaveLine (data, req, res) {
  // Force the user stored in the line.
  data.user = req.user._id

  return sanitizeInput(data)
    .then(data => saveLine(data))
    .then(line => res.json(filterLine(line)))
}

function parseErrors (err) {
  return {
    errors: [{
      fieldName: err.field || 'global',
      message: err.message
    }]
  }
}
