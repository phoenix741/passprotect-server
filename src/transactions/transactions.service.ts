import { Model } from 'mongoose';
import { PubSub } from 'apollo-server-express';
import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { TransactionTypeEnum } from '../shared/interfaces/transaction-type-enum.interface';
import { TransactionEntity } from './models/transaction.entity';
import { IFindAllTransactionParams } from './models/transaction.models';
import { LineEntity } from '../lines/models/line.entity';
import { createHash } from 'crypto';
import { ObjectID, Binary } from 'bson';

export const TRANSACTION_ADDED_TOPIC = 'transactionAdded';

@Injectable()
export class TransactionsService {
  public transactionPubSub = new PubSub();

  constructor(
    @InjectModel('Transaction')
    private readonly transactionModel: Model<TransactionEntity>,
  ) {}

  async findAll(userId: string, params: IFindAllTransactionParams = {}): Promise<TransactionEntity[]> {
    const filter: IFindAllTransactionDto = { user: userId };
    if (params.earliest) {
      filter.updatedAt = { $gte: params.earliest };
    }
    return await this.transactionModel
      .find(filter)
      .sort({ updatedAt: 1 })
      .exec();
  }

  async createTransaction(type: TransactionTypeEnum, before?: LineEntity, after?: LineEntity): Promise<TransactionEntity> {
    const base = after || before;
    if (!base) {
      throw new Error('Transaction: before or after should be defined');
    }

    const transaction: ICreateTransactionDto = {
      _id: new ObjectID(),
      type,
      line: base._id,
      user: base.user,
      before,
      after,
      sha512:
        after &&
        createHash('sha512')
          .update(after.encryption.content instanceof Binary ? after.encryption.content.buffer : after.encryption.content)
          .digest('hex'),
    };

    const createdTransaction = await this.transactionModel.create(transaction);
    this.transactionAdded(createdTransaction);
    return createdTransaction;
  }

  /**
   * Publish a new transaction to applications.
   *
   * @param {Object} transaction Transaction object
   */
  private transactionAdded(transaction: TransactionEntity) {
    this.transactionPubSub.publish(TRANSACTION_ADDED_TOPIC, {
      transactionAdded: transaction,
    });
  }
}

interface IFindAllTransactionDto {
  user: string;
  updatedAt?: {
    $gte: Date;
  };
}

interface ICreateTransactionDto {
  _id: ObjectID;
  type: TransactionTypeEnum;
  line: ObjectID;
  user: string;
  before: object;
  after: object;
  sha512: string;
}
