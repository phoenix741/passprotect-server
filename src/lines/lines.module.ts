import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { LineSchema } from './schemas/line.schema';
import { LinesService } from './lines.service';
import { SharedModule } from '../shared/shared.module';
import { TransactionsModule } from '../transactions/transactions.module';

@Module({
  imports: [SharedModule, MongooseModule.forFeature([{ name: 'Line', schema: LineSchema }]), TransactionsModule],
  providers: [LinesService],
  exports: [LinesService],
})
export class LinesModule {}
