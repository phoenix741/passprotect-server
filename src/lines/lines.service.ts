import { Model } from 'mongoose';
import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { LineToCreate, LineToUpdate } from './models/line.models';
import { LineEntity } from './models/line.entity';
import { ObjectID } from 'bson';
import { TransactionsService } from '../transactions/transactions.service';
import { TransactionEntity } from '../transactions/models/transaction.entity';
import { TransactionTypeEnum } from '../shared/interfaces/transaction-type-enum.interface';

interface IFindAllFilter {
  user?: string;
}

@Injectable()
export class LinesService {
  constructor(
    @InjectModel('Line') private readonly lineModel: Model<LineEntity>,
    private readonly transactionService: TransactionsService,
  ) {}

  async findAll(userId: string): Promise<LineEntity[]> {
    if (userId) {
      const filter: IFindAllFilter = { user: userId };
      const sort = { group: 1, label: 1 };
      return await this.lineModel
        .find(filter)
        .sort(sort)
        .exec();
    }
    return [];
  }

  async findById(id: ObjectID) {
    return await this.lineModel.findById(id);
  }

  async createLine(line: LineToCreate): Promise<LineEntity> {
    const lineToCreate = new this.lineModel(line);
    const newLine = await lineToCreate.save();
    await this.transactionService.createTransaction(
      TransactionTypeEnum.line,
      null,
      newLine,
    );
    return newLine;
  }

  async updateLine(
    originalLine: LineEntity,
    line: LineToUpdate,
  ): Promise<LineEntity> {
    const { _id, ...cleanLine } = line;

    const oldLine = new this.lineModel(originalLine.toObject());
    originalLine.set(cleanLine);
    const newLine = await originalLine.save();

    await this.transactionService.createTransaction(
      TransactionTypeEnum.line,
      oldLine,
      newLine,
    );
    return newLine;
  }

  async removeLine(id: ObjectID): Promise<TransactionEntity> {
    const oldLine = await this.lineModel.findById(id);
    await this.lineModel.deleteOne({ _id: id });
    return this.transactionService.createTransaction(
      TransactionTypeEnum.line,
      oldLine,
      null,
    );
  }

  async getGroups(userId: string): Promise<string[]> {
    const filter = { user: userId };
    return this.lineModel.distinct('group', filter);
  }
}
