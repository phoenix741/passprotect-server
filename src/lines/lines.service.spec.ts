import { Test } from '@nestjs/testing';
import { getModelToken } from '@nestjs/mongoose';
import { LineEntity } from './models/line.entity';
import * as mongoose from 'mongoose';
import mockingoose from 'mockingoose';
import { Model } from 'mongoose';
import { LineSchema } from './schemas/line.schema';
import { LinesService } from './lines.service';
import { TransactionsService } from '../transactions/transactions.service';
import { ObjectID } from 'bson';
import { LineTypeEnum } from '../shared/interfaces/line-type-enum.interface';
import { TransactionTypeEnum } from '../shared/interfaces/transaction-type-enum.interface';

describe('LinesService', () => {
  let lineModel: Model<LineEntity>;
  let linesService: LinesService;
  let transactionsService;

  const doc = {
    _id: new ObjectID('5ce1ae737786b25be429b81f'),
    type: LineTypeEnum.card,
    label: 'label',
    group: 'group',
    user: 'user',
  };

  const docToCreate = {
    type: LineTypeEnum.card,
    label: 'label',
    group: 'group',
    user: 'user',
    encryption: {
      salt: Buffer.from('salt'),
      iv: Buffer.from('iv'),
      authTag: Buffer.from('authTag'),
      content: Buffer.from('content'),
    },
  };

  beforeEach(async () => {
    lineModel = mongoose.model('Line', LineSchema);
    transactionsService = {
      createTransaction: jest.fn(),
    };

    const module = await Test.createTestingModule({
      providers: [
        LinesService,
        {
          provide: getModelToken('Line'),
          useValue: lineModel,
        },
        {
          provide: TransactionsService,
          useValue: transactionsService,
        },
      ],
    }).compile();

    linesService = module.get<LinesService>(LinesService);
  });

  describe('findAll', () => {
    it('call find all - with userId', async () => {
      const userId = 'userId';
      mockingoose(lineModel).toReturn([doc], 'find');
      const lines = await linesService.findAll(userId);
      expect(lines.map(line => line.toObject())).toMatchSnapshot(
        'result of find all with userId',
      );
    });

    it('call find all - without userId', async () => {
      const userId = '';
      mockingoose(lineModel).toReturn([doc], 'find');
      const lines = await linesService.findAll(userId);
      expect(lines).toMatchSnapshot('result of find all without userId');
    });
  });

  describe('findById', () => {
    it('call find by id', async () => {
      mockingoose(lineModel).toReturn(doc, 'findOne');
      const line = await linesService.findById(doc._id);
      expect(line.toObject()).toMatchSnapshot('result of find by id');
    });
  });

  describe('createLine', () => {
    it('call create line - no error', async () => {
      mockingoose(lineModel).toReturn(doc, 'save');
      const line = await linesService.createLine(docToCreate);
      expect(transactionsService.createTransaction).toHaveBeenCalledWith(
        TransactionTypeEnum.line,
        null,
        expect.objectContaining(doc),
      );
      expect(line.toObject()).toMatchSnapshot(
        'result after the creation of the line',
      );
    });

    it('call create line - with error', async () => {
      mockingoose(lineModel).toReturn(new Error('My Error'), 'save');
      await expect(linesService.createLine(docToCreate)).rejects.toThrow(
        'My Error',
      );
      expect(transactionsService.createTransaction).not.toHaveBeenCalledWith(
        TransactionTypeEnum.line,
        null,
        expect.objectContaining(doc),
      );
    });
  });

  describe('updateLine', () => {
    it('call update line - no error', async () => {
      mockingoose(lineModel).toReturn(doc, 'findOne');
      const oldLine = await linesService.findById(doc._id);

      mockingoose(lineModel).toReturn(doc, 'save');
      const line = await linesService.updateLine(oldLine, docToCreate);
      expect(transactionsService.createTransaction).toHaveBeenCalledWith(
        TransactionTypeEnum.line,
        expect.objectContaining(doc),
        expect.objectContaining(doc),
      );
      expect(line.toObject()).toMatchSnapshot('result after updating the line');
    });

    it('call update line - with error', async () => {
      mockingoose(lineModel).toReturn(doc, 'findOne');
      const oldLine = await linesService.findById(doc._id);

      mockingoose(lineModel).toReturn(new Error('My Error'), 'save');
      await expect(
        linesService.updateLine(oldLine, docToCreate),
      ).rejects.toThrow('My Error');
      expect(transactionsService.createTransaction).not.toHaveBeenCalledWith(
        TransactionTypeEnum.line,
        expect.objectContaining(doc),
        expect.objectContaining(doc),
      );
    });
  });

  describe('removeLine', () => {
    it('call remove line - no error', async () => {
      mockingoose(lineModel).toReturn(doc, 'deleteOne');
      await linesService.removeLine(doc._id);
      expect(transactionsService.createTransaction).toHaveBeenCalledWith(
        TransactionTypeEnum.line,
        expect.objectContaining(doc),
        null,
      );
    });

    it('call remove line - with error', async () => {
      mockingoose(lineModel).toReturn(new Error('My Error'), 'deleteOne');
      await expect(linesService.removeLine(doc._id)).rejects.toThrow(
        'My Error',
      );
      expect(transactionsService.createTransaction).not.toHaveBeenCalledWith(
        TransactionTypeEnum.line,
        expect.objectContaining(doc),
        null,
      );
    });
  });

  describe('getGroups', () => {
    it('call get groups', async () => {
      const groups = ['group1', 'group2', 'group3'];
      mockingoose(lineModel).toReturn(groups, 'distinct');
      expect(await linesService.getGroups('userId')).toMatchSnapshot(
        'result for getting groups',
      );
    });
  });
});
