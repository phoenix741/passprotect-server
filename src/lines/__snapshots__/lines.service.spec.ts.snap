// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinesService createLine call create line - no error: result after the creation of the line 1`] = `
Object {
  "_id": "5ce1ae737786b25be429b81f",
  "group": "group",
  "label": "label",
  "type": "card",
  "user": "user",
}
`;

exports[`LinesService findAll call find all - with userId: result of find all with userId 1`] = `
Array [
  Object {
    "_id": "5ce1ae737786b25be429b81f",
    "group": "group",
    "label": "label",
    "type": "card",
    "user": "user",
  },
]
`;

exports[`LinesService findAll call find all - without userId: result of find all without userId 1`] = `Array []`;

exports[`LinesService findById call find by id: result of find by id 1`] = `
Object {
  "_id": "5ce1ae737786b25be429b81f",
  "group": "group",
  "label": "label",
  "type": "card",
  "user": "user",
}
`;

exports[`LinesService getGroups call get groups: result for getting groups 1`] = `
Array [
  "group1",
  "group2",
  "group3",
]
`;

exports[`LinesService updateLine call update line - no error: result after updating the line 1`] = `
Object {
  "_id": "5ce1ae737786b25be429b81f",
  "group": "group",
  "label": "label",
  "type": "card",
  "user": "user",
}
`;
